langworld_db_data/adders/listed_value_adder.py

```diff
@@ -16,54 +16,98 @@ def add_listed_value(
         self,
         feature_id: str,
         new_value_en: str,
         new_value_ru: str,
+        index_to_assign: int,
     ) -> str:
+        """
+        Returns ID of new value and tuple of indices that must be incremented in feature profiles.
+        index_to_assign means number that must be assigned to the new value within the feature, ex. 13 for A-2-13.
+        If no index_to_assign is given, the new value will be added as the last one in the feature.
+        index_to_assign must be greater than 0.
+        """
+
         rows = read_dicts_from_csv(self.input_file_with_listed_values)
 
         if not [r for r in rows if r["feature_id"] == feature_id]:
             raise ListedValueAdderError(f"Feature ID {feature_id} not found")
 
-        index_of_last_row_for_given_feature = 0
-        id_of_new_value = ""
+        # Collect all indices and line numbers of given feature into value_indices_to_inventory_line_numbers
+        value_indices_to_inventory_line_numbers = (
+            self._get_indices_and_their_line_numbers_in_features_listed_values(
+                rows=rows,
+                feature_id=feature_id,
+                new_value_en=new_value_en,
+                new_value_ru=new_value_ru,
+            )
+        )
 
-        for i, row in enumerate(rows):
-            if row["feature_id"] != feature_id:
-                continue
+        # Check if index_to_assign is -1 or belongs to range of indices in value_indices_to_inventory_line_numbers
+        last_index_in_feature = value_indices_to_inventory_line_numbers[-1]["index"]
+        acceptable_indices_to_assign = [-1] + [i for i in range(1, last_index_in_feature)]
+        if index_to_assign not in acceptable_indices_to_assign:
+            raise ValueError(
+                f"Invalid index_to assign (must be between 1 and {last_index_in_feature}, "
+                f"{index_to_assign} was given)"
+            )
+
+        # id_of_new_value and line_number_of_new_value are assigned for default case and then changed if necessary
+        id_of_new_value = (
+            f"{feature_id}{ID_SEPARATOR}{value_indices_to_inventory_line_numbers[-1]['index'] + 1}"
+        )
+        line_number_of_new_value = value_indices_to_inventory_line_numbers[-1]["line number"] + 1
 
-            if row["en"] == new_value_en or row["ru"] == new_value_ru:
-                raise ListedValueAdderError(
-                    f"Row {row} already contains value you are trying to add"
-                )
+        # If value is inserted into range of values, IDs following it must be incremented
+
+        if index_to_assign > -1:
+
+            id_of_new_value = f"{feature_id}{ID_SEPARATOR}{index_to_assign}"
+
+            # Go through values of the feature, ignore indices less than index_to_assign,
+            # increment all other indices
+            rows = self._increment_ids_whose_indices_are_not_less_than_index_to_assign_in_rows(
+                rows=rows,
```

Comment:
hm... assigning the result of function to the same mutable variable doesn't look like a good idea

Reply:
Now the result is called `rows_with_updated_value_indices`, however I also had to introduce this variable to the `if index_to_assign in (-1, last_index_in_feature + 1)` block, just by copying `rows` content